### ISSUE TITLE
Translate Plan Defninition Task content 

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#3301]
**Describe the feature request.**
Add a new translation service in the rules engine to adapt to the selected language by user 

**Additional context**
https://github.com/ona-health/sid-dto-fhir-resources/issues/877

**Comment by f-odhiambo** [comment]
Loud thinking: We could add a new translation service in the rules engine to adapt to the selected language by the user.
So thinking along the lines of

1. From the PD extract tasks with maybe multiple description strings (we need to figure out how to store that)
2. In the Register or profile get the locale (language) in use then on the rules, select the appropriate string based on the ISO language code 
3. Add text to the placeholder in the card view or text placeholder  
4. On refresh repeat step 2

**Comment by ndegwamartin** [comment]
This is how the FHIR SDK handles translations - https://github.com/opensrp/fhircore/blob/main/android/engine/src/test/assets/sample_patient_registration.json#L81-L98

```json
              "text": "First Name",
              "_text": {
                "extension": [
                  {
                    "url": "http://hl7.org/fhir/StructureDefinition/translation",
                    "extension": [
                      {
                        "url": "lang",
                        "valueCode": "sw"
                      },
                      {
                        "url": "content",
                        "valueString": "Jina la kwanza"
                      }
                    ]
                  }
                ]
              },
```

So basically the translation extension is used and is wrapped around an underscore prefixed field that matches the translatable field (**text** in this case). This is inline with the FHIR spec for localization - https://build.fhir.org/languages.html#ext

We could follow the same approach e.g. for tasks such that `Task.description` has a corresponding T`ask._description` which will will contain the translation extensions.

**Comment by pld** [comment]
This looks like the correct approach, is the issue that the rendering engine doesn't understand to look for the translations if they exist?

**Comment by f-odhiambo** [comment]
NB. 

The following texts are also not being translated on the current main 
1. Actions on Paginated forms 
2. Error prompts for validated/required fields 

CC @Lentumunai-Mark  @ndegwamartin

**Comment by f-odhiambo** [comment]
**Is your feature request related to a problem? Please describe.**
CQL evaluation results translated post computation based on the selected app language preference 

**Describe the solution you'd like**
The current CQL statements do not have translation for the evaluations that are text based post computation

**Subtask**
- [ ] Add language translation for G6PD test and G6PD control mode questionnaires
- [ ] Research on how to do CQL results language conversion 

**Additional context**
N/A

**Screenshots**

<img src="https://user-images.githubusercontent.com/4540684/150754212-9ee907ba-c87a-4533-9c97-bd4e1ed04919.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150754219-6e8af6c6-f80e-4b41-8444-0f664fa69681.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150754226-b14ffa70-86ec-4965-b6a2-7206e7578b68.png" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/4540684/150754230-dc250555-ed5b-4f6f-bb71-14eb971d8549.png" width="200" height="400" />

**Comment by FikriMilano** [comment]
> This looks like the correct approach, is the issue that the rendering engine doesn't understand to look for the translations if they exist?

@pld yea, I think the idea is to have rules engine process the translated text right after getting the resources to be displayed in a config.

So, by the time those resources are used, the texts are already translated.

That is if we are to use the _text translation approach (which i think it's better). This way, we don't need to use custom tags.
@f-odhiambo

**Comment by FikriMilano** [comment]
> NB. 
> 
> The following texts are also not being translated on the current main 
> 1. Actions on Paginated forms 
> 2. Error prompts for validated/required fields  

Actually, these are translatable by using strings.xml

Stuff like `Next`, `Previous`, etc.

**Comment by f-odhiambo** [comment]
> > NB.
> > The following texts are also not being translated on the current main
> > 
> > 1. Actions on Paginated forms
> > 2. Error prompts for validated/required fields
> 
> Actually, these are translatable by using strings.xml
> 
> Stuff like `Next`, `Previous`, etc.

@FikriMilano this is broken on the `main` branch at the moment  CC @ndegwamartin

**Comment by pld** [comment]
> Another use case
> 
> **Is your feature request related to a problem? Please describe.** CQL evaluation results translated post computation based on the selected app language preference
> 
> **Describe the solution you'd like** The current CQL statements do not have translation for the evaluations that are text based post computation
> 
> **Subtask**
> 
>     * [ ]  Add language translation for G6PD test and G6PD control mode questionnaires
> 
>     * [ ]  Research on how to do CQL results language conversion
> 
> 
> **Additional context** N/A
> 
> **Screenshots**
> 
> <img alt="" width="200" height="400" src="https://user-images.githubusercontent.com/4540684/150754212-9ee907ba-c87a-4533-9c97-bd4e1ed04919.png">
> 
> <img alt="" width="200" height="400" src="https://user-images.githubusercontent.com/4540684/150754219-6e8af6c6-f80e-4b41-8444-0f664fa69681.png">
> 
> <img alt="" width="200" height="400" src="https://user-images.githubusercontent.com/4540684/150754226-b14ffa70-86ec-4965-b6a2-7206e7578b68.png">
> 
> <img alt="" width="200" height="400" src="https://user-images.githubusercontent.com/4540684/150754230-dc250555-ed5b-4f6f-bb71-14eb971d8549.png">

OK, this can be covered by a different issue since it's not as high priority

**Comment by pld** [comment]
@FikriMilano can you please rewrite the description to focus only on what is needed for the Engage project you are working on? Let's solve those first, exhausting the options with configs before we make product changes, thank you

**Comment by FikriMilano** [comment]
@pld updated

**Comment by FikriMilano** [comment]
> This looks like the correct approach, is the issue that the rendering engine doesn't understand to look for the translations if they exist?

@pld yeah, the rules engine/config doesn't have the implementation to retrieve translated text from translation extension.

Need to do that implementation where at the instance that property is called in anywhere within the code/triggered by rules engine, we retrieve the right translation according to the device language during that time point.

We can re-use this function from SDK, then call it on anywhere within our code.
https://github.com/google/android-fhir/blob/master/common/src/main/java/com/google/android/fhir/MoreTypes.kt#L97-L99

**Comment by pld** [comment]
OK got it, thank you that is very clear! My understanding is whether we do interpolated codes (option 1) or _text (option 2), we will need to add code to fhircore to respectively understand the templated data or parse the fhir resource extension content, is that correct?

If that's true, as @ndegwamartin noted we prefer (option 2). Do you feel comfortable starting to implement that? I'm happy to discuss the implementation plan

@ellykits also please add any thoughts you have on the approach, potential implementation plan

thank you everyone

**Comment by FikriMilano** [comment]
@pld yeah, happy w option 2: it's FHIR based and the translated text is more straightforward to retrieve (no need to have multiple language files)

# Implementation Plan (Point 2 and 3)
1. Hardcoded - when StringType properties are directly accessed anywhere within the codebase, they should use getLocalizedText function
2. Config rules - allow getLocalizedText function to be accessed via rules, so any StringType property can use it

# Implementation Plan (Point 1)
Implementing 1 of these should suffice:
1. Pass a reserved variable called `%language` to the FHIRPath context, the translation logic on which language to pick, will be written in the questionnaire it self (FHIRPath context is like a place that has list of variables that FHIRPath can use)
2. Pass the getLocalizedText function to FHIRPath so it can be used by the expressions (not sure if FHIRPath allows this)
3. Use `_expression` + translation extension, and upon FHIRPath calculation, the code should check what's the current device language, pick the right translation extension, then get the right expression, then do the calculation (had a look in the past, I doubt there's such thing as `_expression`, but it's reasonable for it to exists)

**Comment by ellykits** [comment]
I like the idea of using extensions for translations and following the FHIR specs. 



> ```json
>               "text": "First Name",
>               "_text": {
>                 "extension": [
>                   {
>                     "url": "http://hl7.org/fhir/StructureDefinition/translation",
>                     "extension": [
>                       {
>                         "url": "lang",
>                         "valueCode": "sw"
>                       },
>                       {
>                         "url": "content",
>                         "valueString": "Jina la kwanza"
>                       }
>                     ]
>                   }
>                 ]
>               },
> ```

We can use this to enhance the translation functionality already in OpenSRP 2. The `RulesEngineService` has an implementation of the `translate` method that should be used to translate string content generated via rules execution including data extracted via FHIRPath. It has a few limitations though and it is opinionated. Here is how it works.

Assuming the text to be translated is `Vaccine status`, the [RulesEngine#translate](https://github.com/opensrp/fhircore/blob/b97dd6607629167b45b2f4c6a5d28978b6e7ad9f/android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt#L191) will attempt to locate `vaccine.status` property in the relevant properties file for the current Locale. It's opinionated in that the keys are expected to be written in the dot notation for translations to work. 

To use the function in rules 

```kotlin
  service.translate('Vaccine status')
```

The implementation is however not suitable for lengthy strings. To avoid the limitation we could use `md5` algorithm to hash the string to get consistent keys as used in the [efcity translation tool](https://github.com/onaio/fhir-tooling/tree/main/efsity-cli#localization). 

As for the enhancement to use extensions, we can approach this in two ways (or both):

1. Translate the resource in the methods that trigger the rules ([RulesFactory#fireRules](https://github.com/opensrp/fhircore/blob/b97dd6607629167b45b2f4c6a5d28978b6e7ad9f/android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt#L104) and [ConfigRulesExecutor#fireRules](https://github.com/opensrp/fhircore/blob/b97dd6607629167b45b2f4c6a5d28978b6e7ad9f/android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/ConfigRulesExecutor.kt#L56)
2. Implement a translate service method in the rules engine service class with the following signature then extract data from the returned resource. 

```kotlin
fun translate(resource: Resource) : Resource
```

If the intention is to translate the text rendered on the screen then I believe running the efcity translation tool should suffice. The SDK should be able to handle Questionnaire translations.
